### PR TITLE
fix(discovery): Fixes aliased search keys + issue shortname + project name queries

### PR DESCRIPTION
**Pull request by rjo100**
Converts issue shortnames and project names where appropriate. Also fixes `issue` and `project` aliased search keys.

No tests currently on snuba `query` executor, do we want to add them?

Fixes SENTRY-RXR

### CODE REVIEW

**Comment by barkbarkimashark** [comment]
Hmm, took a look at the referenced issue and it looks like someone was trying to make a query like `project:react-native` which does seem to have a UI dropdown hint:
Screen Shot 2023-01-30 at 7 35 37 PM

which is kinda ambiguous. Presumably, the user was trying to search for project slug or name? I wonder what `project` filter term is supposed to actually search on? I did a cursory look at some events on that issue and its kinda all over the place. Some users seem to have values like `cln_dkv_<redacted>_29c4_chatbot_es`, and some have stuff like `<redacted>-API`. 

This issue might be due to our UI showing that `project` filter hint and users accidentally using that filter without knowing what it's supposed to do.

**Comment by markstory** [comment]
> Presumably, the user was trying to search for project slug or name? I wonder what project filter term is supposed to actually search on?

Historically the `project` key has been an alias for the sentry project. Customers often put a project tag on their events which can map to cloud hosting projects. To search this key we have a `tags[project]` alias.

**Comment by rjo100** [comment]
So there's basically 3 different bugs here:
1. All key aliasing (as best I can tell) is broken. The change in `src/sentry/search/events/filter.py` fixes that
2. The `issue` key alias was broken because we weren't translating the shortnames into ids
3. The `project` key alias was broken because we weren't passing the project ids appropriately to the query

**Comment by rjo100** [comment]
I found one or two more changes I want to make.

**Comment by barkbarkimashark** [comment]
I think its worthwhile to add some tests here: https://github.com/getsentry/sentry/blob/2683b1f05d79e4d84f364aece48a2118a300f674/tests/snuba/search/test_backend.py

to verify search queries using the key aliases: `project` and `issue` work.

File: src/sentry/api/serializers/models/group.py [hunk 1]
```diff
@@ -900,22 +900,32 @@ def __init__(
...
+                        else convert_search_filter_to_snuba_query(
+                            search_filter,
+                            params={
+                                "organization_id": organization_id,
+                                "project_id": project_ids,
+                                "environment_id": environment_ids,
+                            },
+                        )
+                    )
+        self.conditions = conditions
```
Comment (barkbarkimashark) [inline src/sentry/api/serializers/models/group.py:928]: I wonder if its worthwhile to filter out conditions like `[None, "=", "value"]` that couldn't resolve the search key alias, or at least throw an error that bubbles up to the front-end.

File: src/sentry/search/events/filter.py [hunk 2]
```diff
@@ -702,250 +631,6 @@ def convert_search_filter_to_snuba_query(
...
-            having = None
-        elif len(args) == 1:
-            having = args[0]
-        else:
-            having = [operator, args]
-
-    return condition, having, projects_to_filter, group_ids
-
-
-def get_filter(query=None, params=None, parser_config_overrides=None):
```
Comment (barkbarkimashark) [inline src/sentry/search/events/filter.py:845]: Woah, I feel like removing this should be in a separate PR ...

Comment (rjo100) [reply src/sentry/search/events/filter.py:845]: Mmm yeah fair

**Comment by rjo100** [comment]
Both the `issue` and `project` keys are undocumented and throw up a warning when trying to search on them. Regardless, we're getting `issue` key searches for sure and sounds like `project` too. Will cut another PR/task to update the docs/warning tooltip

File: src/sentry/api/serializers/models/group.py (hunk 1 above)
Comment (rjo100) [reply src/sentry/api/serializers/models/group.py:927]: I mean it just says "Request failed" basically. It's not actionable from a user perspective and we are logging + creating an issue from it. Feel like we can leave as-is

**Comment by rjo100** [comment]
Still want to add a couple tests + comments.

File: src/sentry/api/serializers/models/group.py (hunk 1 above)
Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/group.py:927]: True. Maybe we could throw a specific error when we transform the `SearchFilter` -> `Condition` and the resulting condition has a `None` key.name. That way the user can fix their query instead of just getting a request failed and not know what's going on?

Comment (rjo100) [reply src/sentry/api/serializers/models/group.py:927]: eh i mean it's not really user actionable because we're converting the name to `None`. it should be that we throw the error like now IMO

Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/group.py:928]: Currently, when searching with a query like `issue:DOESNT-EXIST` the front-end shows an internal error. 
Screen Shot 2023-02-02 at 3 07 37 PM

Your PR should fix the cases for issue and project querying but any future keys that don't transform from SearchFilter -> Condition will result in a non-helpful generic Internal error. If we do a check after the search filter transformation for any conditions that have a `None` key, we could explicitly raise an error that would show up in the front-end so at least the user has a better idea why the query failed.

Comment (rjo100) [reply src/sentry/api/serializers/models/group.py:928]: Right I'm just saying that this really is an internal error and there's nothing the user can do about it. I don't think we really want to add a message like "we have a bug"

Comment (barkbarkimashark) [reply src/sentry/api/serializers/models/group.py:928]: Yeah, fair enough.

File: src/sentry/search/snuba/executors.py [hunk 3]
```diff
@@ -156,8 +156,19 @@ def _convert_search_filters(
...
+                    "project_id": project_ids,
+                    "environment": environments,
+                },
+            )
+            # if no re-formatted conditions, use fallback method
             converted_filters.append(
-                convert_search_filter_to_snuba_query(
+                conditions[0]
+                if conditions
+                else convert_search_filter_to_snuba_query(
```
Comment (wedamija) [inline src/sentry/search/snuba/executors.py:171]: Kind of wonder if this `try format_search_filter else convert_search_filter_to_snuba_query` logic should be encapsulated somewhere. We only do it in two places though so maybe it's fine.

**Review by wedamija — APPROVED** [review]
This looks good to me. I'm always a bit paranoid about changes to search, so let's be a bit cautious with this and follow https://www.notion.so/sentry/RFC-Ensuring-Safe-Deployments-39ab017e4e914e2a9fa427a456302e0a

**Comment by rjo100** [comment]
agreed on the rollout strategy

File: src/sentry/search/snuba/executors.py (hunk 3 above)
Comment (rjo100) [reply src/sentry/search/snuba/executors.py:171]: probably but this logic is already kind of supposed to be unified. there's a big comment about it in one of the files. I think in general this entire situation needs a refactor/overhaul considering we have a lot of unused code and it's not clear where all of the query codepaths are
